File: backend/Api/GetUserHandler.cs
using System;
using System.Threading.Tasks;
using Core.UserDomain;

namespace Api
{
    public struct GetUserRequest
    {
        public string Id { get; set; }
    }

    public struct GetUserResponse
    {
        public string Id { get; set; }
        public string Name { get; set; }
    }

    public class GetUserHandler : BaseHandler<GetUserRequest, GetUserResponse>
    {
        private readonly IUserRepository userRepository;

        public GetUserHandler(IUserRepository userRepository)
        {
            this.userRepository = userRepository;
        }

        protected override async Task<GetUserResponse> Handle(GetUserRequest request)
        {
            var user = await userRepository.GetUser(new UserId(Guid.Parse(request.Id)));

            return new GetUserResponse
            {
                Id = user.Id.Value.ToString(),
                Name = user.Name
            };
        }
    }
}

File: packages/backend/src/Api/GetUserHandler.ts
import { User, UserId, UserRepository, UserSerializer } from "@pokki/core";
import { Serializer } from "@tsukiy0/tscore";
import { Handler } from "./Handler";

export class GetUserRequest {
  constructor(public readonly id: UserId) {}
}

export type GetUserRequestJson = {
  id: string;
};

export const GetUserRequestSerializer: Serializer<
  GetUserRequest,
  GetUserRequestJson
> = {
  serialize: (input: GetUserRequest): GetUserRequestJson => {
    return {
      id: input.id.toString(),
    };
  },
  deserialize: (input: GetUserRequestJson): GetUserRequest => {
    return new GetUserRequest(new UserId(input.id));
  },
};

export class GetUserHandler extends Handler<GetUserRequest, User> {
  constructor(public readonly userRepository: UserRepository) {
    super(GetUserRequestSerializer, UserSerializer);
  }

  handle(request: GetUserRequest): Promise<User> {
    return this.userRepository.getUser(request.id);
  }
}
